pkg/ruler/mapper_test.go

```diff
@@ -346,7 +350,87 @@ func Test_mapper_MapRulesMultipleFiles(t *testing.T) {
 		require.False(t, exists)
 		require.NoError(t, err)
 	})
+}
+
+func Test_mapper_MapRulesMultipleTenants(t *testing.T) {
+	l := util_log.MakeLeveledLogger(os.Stdout, "info")
+	setupRuleSets()
+	m := &mapper{
+		Path:   "/rules",
+		FS:     afero.NewMemMapFs(),
+		logger: l,
+	}
+
+	t.Run("basic rulegroup tenant 1", func(t *testing.T) {
+		updated, files, err := m.MapRules(testUser, initialRuleSet)
+		require.True(t, updated)
+		require.Len(t, files, 1)
+		require.Equal(t, fileOnePath, files[0])
+		require.NoError(t, err)
+
+		exists, err := afero.Exists(m.FS, fileOnePath)
+		require.True(t, exists)
+		require.NoError(t, err)
+	})
+
+	t.Run("basic rulegroup tenant 2 still considered new", func(t *testing.T) {
+		updated, files, err := m.MapRules(testUser2, initialRuleSet)
+		require.True(t, updated)
+		require.Len(t, files, 1)
+		require.Equal(t, fileOneUserTwoPath, files[0])
+		require.NoError(t, err)
+
+		exists, err := afero.Exists(m.FS, fileOneUserTwoPath)
+		require.True(t, exists)
+		require.NoError(t, err)
+	})
+
+	t.Run("simultaneous update and add tenant 2", func(t *testing.T) {
+		updated, files, err := m.MapRules(testUser2, twoFilesRuleSet)
+		require.True(t, updated)
+		require.Len(t, files, 2)
+		require.True(t, sliceContains(t, fileOneUserTwoPath, files))
+		require.True(t, sliceContains(t, fileTwoUserTwoPath, files))
+		require.NoError(t, err)
 
+		exists, err := afero.Exists(m.FS, fileOnePath)
+		require.True(t, exists)
+		require.NoError(t, err)
+		exists, err = afero.Exists(m.FS, fileOneUserTwoPath)
+		require.True(t, exists)
+		require.NoError(t, err)
+		exists, err = afero.Exists(m.FS, fileTwoUserTwoPath)
+		require.True(t, exists)
+		require.NoError(t, err)
+	})
+
+	t.Run("identical rulegroup tenant 1 not considered updated", func(t *testing.T) {
+		updated, files, err := m.MapRules(testUser, initialRuleSet)
+		require.False(t, updated)
+		require.Len(t, files, 1)
+		require.NoError(t, err)
+
+		exists, err := afero.Exists(m.FS, fileOnePath)
+		require.True(t, exists)
+		require.NoError(t, err)
+	})
+
+	t.Run("removal of tenant 1 groups keeps tenant 2 groups", func(t *testing.T) {
+		updated, files, err := m.MapRules(testUser, map[string][]rulefmt.RuleGroup{})
+		require.True(t, updated)
+		require.Len(t, files, 0)
+		require.NoError(t, err)
+
+		exists, err := afero.Exists(m.FS, fileOnePath)
```

Comment:
The "exists" assertion seems like a good candidate for a helper method since it's called so many times.